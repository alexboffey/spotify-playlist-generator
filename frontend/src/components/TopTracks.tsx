import React, { useState } from "react";
import { Query } from "react-apollo";
import { Avatar, message, List, Form, Select } from "antd";

import Header from "./Header";
import {
  IMyTopTracksQuery,
  MY_TOP_TRACKS_QUERY
} from "../graphql/myTopTracksQuery";

const TopTracks: React.FunctionComponent = () => {
  const initialTimeRange = "short_term";
  const [timeRange, setTimeRange] = useState<string>(initialTimeRange);

  return (
    <Query<IMyTopTracksQuery>
      query={MY_TOP_TRACKS_QUERY}
      variables={{ limit: 20, time_range: timeRange }}
      notifyOnNetworkStatusChange
    >
      {({ data, loading, error, refetch }) => {
        if (error) return message.error(error);

        return (
          <React.Fragment>
            <Header
              title="My Top Tracks"
              action={() => (
                <Form layout="inline">
                  <Form.Item label="Time Range">
                    <Select
                      defaultValue={initialTimeRange}
                      onChange={(value: string) => {
                        setTimeRange(value);
                        refetch();
                      }}
                    >
                      <Select.Option value="short_term">
                        Short Term
                      </Select.Option>
                      <Select.Option value="medium_term">
                        Medium Term
                      </Select.Option>
                      <Select.Option value="long_term">Long Term</Select.Option>
                    </Select>
                  </Form.Item>
                </Form>
              )}
            />

            <List loading={loading}>
              {data &&
                !loading &&
                data.myTopTracks.items.map(({ name, id, album, artists }) => (
                  <List.Item key={id}>
                    <List.Item.Meta
                      avatar={
                        <Avatar
                          shape="square"
                          size="large"
                          src={album.images[0] ? album.images[0].url : ""}
                        >
                          {name.split("")[0]}
                        </Avatar>
                      }
                      title={<span>{name}</span>}
                      description={artists.map(({ name }) => name).join(", ")}
                    />
                  </List.Item>
                ))}
            </List>
          </React.Fragment>
        );
      }}
    </Query>
  );
};

export default TopTracks;
