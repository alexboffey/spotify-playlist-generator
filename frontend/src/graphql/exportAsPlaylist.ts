import gql from "graphql-tag";

export const EXPORT_AS_PLAYLIST_MUTATION = gql`
  mutation EXPORT_AS_PLAYLIST_MUTATION(
    $playlistName: String
    $tracks: [String]!
  ) {
    exportAsPlaylist(playlistName: $playlistName, tracks: $tracks) {
      name
    }
  }
`;

export interface IExportAsPlaylistMutation {
  exportAsPlaylist: {
    name: string;
  };
}
