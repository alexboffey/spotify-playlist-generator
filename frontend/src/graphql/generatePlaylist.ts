import gql from "graphql-tag";

export const GENERATE_PLAYLIST_QUERY = gql`
  query GENERATE_PLAYLIST_QUERY($seeds: String) {
    generatePlaylist(seeds: $seeds) {
      tracks {
        id
        name
        preview_url
        artists {
          id
          name
        }
        album {
          id
          name
          images {
            url
          }
        }
      }
    }
  }
`;

export interface IGeneratePlaylistQuery {
  generatePlaylist: {
    tracks: Array<{
      id: string;
      name: string;
      preview_url?: string;
      artists: Array<{ id: string; name: string }>;
      album: { id: string; name: string; images: Array<{ url: string }> };
    }>;
  };
}
