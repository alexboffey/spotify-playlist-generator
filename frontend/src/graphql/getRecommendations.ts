import gql from "graphql-tag";

export const GET_RECOMMENDATIONS_QUERY = gql`
  query GET_RECOMMENDATIONS_QUERY($seed_artists: String!) {
    getRecommendations(seed_artists: $seed_artists) {
      tracks {
        id
        name
        artists {
          id
          name
        }
        album {
          id
          name
        }
      }
    }
  }
`;

export interface IGetRecommendationsQuery {
  getRecommendations: {
    tracks: Array<{
      id: string;
      name: string;
      artists: { id: string; name: string };
      album: { id: string; name: string };
    }>;
  };
}
