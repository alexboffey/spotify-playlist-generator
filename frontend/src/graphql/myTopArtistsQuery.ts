import gql from "graphql-tag";

export const MY_TOP_ARTISTS_QUERY = gql`
  query MY_TOP_ARTISTS_QUERY(
    $limit: Int = 20
    $time_range: TimeRange = long_term
  ) {
    myTopArtists(limit: $limit, time_range: $time_range) {
      items {
        id
      }
    }
  }
`;

export interface IMyTopArtistsQuery {
  myTopArtists: {
    items: Array<{ id: string }>;
  };
}
