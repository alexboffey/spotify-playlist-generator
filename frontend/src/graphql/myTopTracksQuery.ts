import gql from "graphql-tag";

export const MY_TOP_TRACKS_QUERY = gql`
  query {
    myTopTracks(limit: 5, time_range: short_term) {
      items {
        name
        id
        href
        album {
          name
        }
      }
      next
      previous
      total
      limit
      href
    }
  }
`;

export interface MyTopTracksData {
  myTopTracks: {
    items: Array<{ name: string; id: string; href: string }>;
    next: string;
    previous: string;
    total: number;
    limit: number;
    href: string;
  };
}
