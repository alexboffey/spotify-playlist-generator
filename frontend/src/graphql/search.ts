import gql from "graphql-tag";

export const SEARCH_ARTISTS_QUERY = gql`
  query SEARCH_ARTISTS_QUERY(
    $query: String!
    $type: SearchType = artist
    $limit: Int
  ) {
    search(query: $query, type: $type, limit: $limit) {
      artists {
        items {
          name
        }
      }
    }
  }
`;

export const SEARCH_TRACKS_QUERY = gql`
  query SEARCH_ARTISTS_QUERY(
    $query: String!
    $type: SearchType = track
    $limit: Int
  ) {
    search(query: $query, type: $type, limit: $limit) {
      tracks {
        items {
          name
        }
      }
    }
  }
`;

export interface ISearchQuery {
  search: string;
  type: string;
}
