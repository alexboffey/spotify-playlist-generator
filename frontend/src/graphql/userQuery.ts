import gql from "graphql-tag"

export const USER_QUERY = gql`
  query {
    me {
      id
      name
      email
      spotifyId
    }
  }
`;

export interface IUserQuery {
  me: {
    id: string;
    name: string;
    email: string;
    spotifyId: string;
  };
}
