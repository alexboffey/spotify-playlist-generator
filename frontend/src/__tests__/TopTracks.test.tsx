import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";

import { MY_TOP_TRACKS_QUERY } from "../graphql/myTopTracksQuery";
import TopTracks from "../components/TopTracks";

const mocks = [
  {
    request: {
      query: MY_TOP_TRACKS_QUERY,
      variables: { limit: 20, time_range: "short_term" }
    },
    result: {
      data: {
        myTopTracks: {
          __typename: "test",
          items: [
            {
              id: "someId",
              name: "someTrackName",
              __typename: "test",
              artists: [
                {
                  id: "someArtistId",
                  name: "someArtistName",
                  __typename: "test"
                }
              ],
              album: {
                id: "someAlbumId",
                name: "someAlbumName",
                images: [
                  {
                    url: "someImageUrl",
                    __typename: "test"
                  }
                ],
                __typename: "test"
              }
            }
          ]
        }
      }
    }
  }
];

describe("<TopArtists />", () => {
  it("Queries, renders data and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TopTracks />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    // Data assertions
    expect(
      wrapper
        .find(".ant-list-item-meta-title")
        .first()
        .text()
    ).toBe("someTrackName");

    // Snapshot
    expect(
      toJSON(wrapper.find("div.ant-list.ant-list-split"))
    ).toMatchSnapshot();
  });
});
