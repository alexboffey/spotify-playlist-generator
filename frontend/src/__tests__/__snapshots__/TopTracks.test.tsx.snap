// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopArtists /> Queries, renders data and matches snapshot 1`] = `
<div
  className="ant-list ant-list-split"
>
  <Spin
    size="default"
    spinning={false}
    wrapperClassName=""
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
        key="container"
      >
        <Item
          key="someId"
        >
          <li
            className="ant-list-item"
          >
            <Meta
              avatar={
                <Avatar
                  shape="square"
                  size="large"
                  src="someImageUrl"
                >
                  s
                </Avatar>
              }
              description="someArtistName"
              title="someTrackName"
            >
              <div
                className="ant-list-item-meta"
              >
                <div
                  className="ant-list-item-meta-avatar"
                >
                  <Avatar
                    shape="square"
                    size="large"
                    src="someImageUrl"
                  >
                    <span
                      className="ant-avatar ant-avatar-lg ant-avatar-square ant-avatar-image"
                      style={Object {}}
                    >
                      <img
                        onError={[Function]}
                        src="someImageUrl"
                      />
                    </span>
                  </Avatar>
                </div>
                <div
                  className="ant-list-item-meta-content"
                >
                  <h4
                    className="ant-list-item-meta-title"
                  >
                    someTrackName
                  </h4>
                  <div
                    className="ant-list-item-meta-description"
                  >
                    someArtistName
                  </div>
                </div>
              </div>
            </Meta>
          </li>
        </Item>
      </div>
    </div>
  </Spin>
</div>
`;
